### PR TITLE
fuzz: add fuzzer for mulmod

### PR DESCRIPTION
**Pull request by holiman**
Fuzzer for mulmod

### CODE REVIEW

File: fuzz.go [hunk 1]
```diff
@@ -120,14 +196,46 @@ func Fuzz(data []byte) int {
...
+func bigMulMod(b1, b2, b3, b4 *big.Int) *big.Int {
+	return b1.Mod(big.NewInt(0).Mul(b2, b3), b4)
+}
+
+func intMulMod(f1, f2, f3, f4 *Int) *Int {
+	return f1.MulMod(f2, f3, f4)
+}
+
+func Fuzz3Args(data []byte) int {
+	if len(data) != 1+32+32+23 {
```
Comment (chfast) [inline fuzz.go:221]:
```suggestion
	if len(data) != 1+32+32+32 {
```

File: fuzz.go [hunk 2]
```diff
@@ -120,14 +196,46 @@ func Fuzz(data []byte) int {
...
 	}
+	op := data[0]
 
-	return 0
+	var x, y, z Int
+	x.SetBytes(data[1:33])
+	y.SetBytes(data[33:65])
+	z.SetBytes(data[65:])
+
+	switch op {
```
Comment (chfast) [inline fuzz.go:231]: I was actually thinking that maybe instead of switching on the first byte just run all operators in a row.

File: fuzz.go [hunk 3]
```diff
@@ -120,14 +196,46 @@ func Fuzz(data []byte) int {
...
+	case opMulmod:
+		if z.IsZero() {
+			return 0
+		}
+		checkThreeArgOp(intMulMod, bigMulMod, x, y, z)
```
Comment (chfast) [inline fuzz.go:236]: Add also `AddMod`.

File: fuzz.go [hunk 4]
```diff
@@ -2,6 +2,7 @@
 // Copyright 2020 uint256 Authors
 // SPDX-License-Identifier: BSD-3-Clause
 
+//go:build gofuzz
```
Comment (chfast) [inline fuzz.go:5]: What is this?

File: fuzz.go [hunk 5]
```diff
@@ -120,14 +196,46 @@ func Fuzz(data []byte) int {
...
+
+func bigMulMod(b1, b2, b3, b4 *big.Int) *big.Int {
+	return b1.Mod(big.NewInt(0).Mul(b2, b3), b4)
+}
+
+func intMulMod(f1, f2, f3, f4 *Int) *Int {
+	return f1.MulMod(f2, f3, f4)
+}
+
+func Fuzz3Args(data []byte) int {
```
Comment (chfast) [inline fuzz.go:220]: Having separate function confuses go-fuzz. This is not needed, you can switch between binary and ternary operators by the length of the input data.

File: fuzz.go (hunk 2 above)
Comment (holiman) [reply fuzz.go:231]: Yes, good idea

File: fuzz.go (hunk 4 above)
Comment (holiman) [reply fuzz.go:5]: go 1.17 build directives have changed. go formt auto-converts the old style to the new one

File: fuzz.go (hunk 5 above)
Comment (holiman) [reply fuzz.go:220]: Sure, we could do that instead

**Comment by holiman** [comment]
Fixed. I guess the old corpus is moot now though, so we could clear that from CI (dunno how much it matters)

**Review by chfast — APPROVED** [review]

**Pull request merged**